src/Data/List/Relation/Unary/All/Properties.agda

```diff
@@ -551,13 +551,11 @@ applyDownFrom⁺₂ f n Pf = applyDownFrom⁺₁ f n (λ _ → Pf _)
 ------------------------------------------------------------------------
 -- tabulate
 
-tabulate⁺ : ∀ {n} {f : Fin n → A} →
-            (∀ i → P (f i)) → All P (tabulate f)
+tabulate⁺ : ∀ {n} {f} → (∀ i → P (f i)) → All P (tabulate {n = n} f)
```

Comment:
So I appreciate the ideological purity at work here in that you should get the type-checker to infer the maximum amount of information that it can! The problem is that we don't have good UI tools to recover that information for the user. I think these changes significantly increase cognitive load on the user, as there is no mouse-over or one single command to recover what the type of `f` is.

So in summary I would probably prefer these type erasure changes to reverted? Thoughts?

Reply:
Thanks for "ideological purity" ;-)
Agree that there is choice here between redundant recoding of 'types-as-machine-checkable-comments' versus "don't import where you don't have to". And as for UI support, indeed using `C-c C-d` to type an expression is fairly limited in applicability...

The trouble is, `tabulate` shouldn't ever have to say *what* is being tabulated (and so, perhaps, neither should the user ever have to think about it, either; in any mode of use where a lemma such as `tabulate⁺` is being deployed, the user presumably already has an `f` in mind which is being `tabulate`d); as for `tabulate` itself, it (un)fortunately *does* need to know the length `n` of the 'table', so any instinct of mine towards 'purity' indeed led me in such directions.

Reply:
I don't want to make this a hill on which to take an extreme stand, but I will now go onto do precisely that...;-) 

I wonder sometimes about exactly how we make the trade-off between 'cognitive load' (which I do care about, believe me) and 'notational clutter' (which, especially if it can be avoided by avoiding expensive imports, perhaps should be...). 

"Redundant recoding" is usually seen in a positive light in the Green/Blackwell "cognitive dimensions" school of HCI-of-PL, but I do wonder in this instance: why give `f` a type, and not `i`? why go to the trouble here, when in the sections on `applyUpTo` and `applyDownFrom` in this module, there is no such fastidiousness?

Reply:
There are other examples eg `Data.List.Membership.Setoid.Properties`, where a module's `import`s rename a `Setoid`'s  `Carrier` to `A`, only so that it can then be mentioned in the type of a `tabulate`d `f`... etc. 

Maybe `tabulate` *is* special, but I don't really see why...

Reply:
> Thanks for "ideological purity" ;-)

Will you believe me if I say that was honestly meant as a positive thing? :sweat_smile: I am always fascinated by the principles raised by you, most of which have never even crossed my mind.... Apologies if it came across otherwise! 

> Agree that there is choice here between redundant recoding of 'types-as-machine-checkable-comments' versus "don't import where you don't have to". 

So this is saving us an import of `Fin` from `Data.Fin.Base` right? The thing is that that doesn't actually help us in any practical way. The dependencies we care about are module level dependencies, i.e. do we import `Data.Fin.Base` at all. The `using` keyword doesn't speed up type-checking or interface loading in Agda. It's purely to avoid naming conflicts and for a certain amount of documentation purposes. The code is still using `Fin` whether we make it implicit or explicit.

> The trouble is, tabulate shouldn't ever have to say what is being tabulated (and so, perhaps, neither should the user ever have to think about it, either;)

I think you may be coming at this from the perspective of an expert user who knows automatically what something called `tabulate` does! A new user, such as myself a few years back, has no idea what the word "tabulate" means and has to piece it together from the implementation and the type.

> but I do wonder in this instance: why give f a type, and not i?

So we have a fixed set of [naming conventions](https://github.com/agda/agda-stdlib/blob/master/notes/style-guide.md#variables-1) in the library (of which `Fin` should be there as `i`, `j`, `k` but isn't for some reason). These conventions are there for exactly the reason that if the user sees an `i` then they should know what type it is. However, there is obviously no convention for `f` beyond it being some sort of function....

> why go to the trouble here, when in the sections on applyUpTo and applyDownFrom in this module

I think that `applyUpTo` and `applyDownFrom` probably _should_ follow this convention, but we've already had bigger fish to fry :cry:

Reply:
Re: purity
I had taken it in a positive vein... I just hadn't quite realised that it came across as 'hardline'

Reply:
Re: 'expert user'

hmmm... I think it's fair to say I'm 'expert', but not necessarily expert as an Agda user, or as a std-lib contributor. 

But even 'novice' users are explicitly invited to use mouse-middle-click to examine definitions... so I'd expect them to do so if they wanted to understand tabulate... which they really should perhaps want to do before examinining its properties here...?

Reply:
Re: naming conventions

Again a fair cop, except as above: f is only the thing-tabulated... so to know anything about f is only to want to know how tabulate works...

.. ie naming conventions giving unambiguous cues as to typing will get you only so far; but the very context-sensitivity of typing (as opposed to naming) is what allows/would/should allow     (the types of) names to be unambiguously interpreted/understood... one of the payoffs of uniqueness of typing?

Reply:
> Other commits in this PR do achieve the aim of eliminating imports completely, but I agree this one does not (might it? Not sure, those zero/suc constructors seem indispensable alas)

I think impossible. Even if we somehow eliminate it syntactically, the module will still transitively depend on it. 

> But even 'novice' users are explicitly invited to use mouse-middle-click to examine definitions... so I'd expect them to do so if they wanted to understand tabulate... which they really should perhaps want to do before examinining its properties here...?

Yes fair enough. But I think every hint we can give people helps. Getting actual documentation is definitely part of the story as well!

> Again a fair cop, except as above: f is only the thing-tabulated... so to know anything about f is only to want to know how tabulate works...

Yes, but doesn't mean that you can't go the other way either, i.e. remember what tabulate does from what `f` is.

To be honest there's also another reason I'm reluctant to make the changes as well. The `{n = n}` notation is seriously ugly :disappointed:

Reply:
Pushed.